### PR TITLE
[Bug] Fixes issues in feature pickling and transmission for subgraphs and blocks

### LINKED ISSUE #2135: [Bug] Pickling a subgraph stores the feature of the original graph
## 🐛 Bug

See reproducing code below.

This affects (1) pickling a graph as a file ~~, and (2) multiprocess neighborhood sampling.~~

## To Reproduce

```python
import dgl
import torch
import pickle
g = dgl.graph(([], []), num_nodes=10000000)
g.ndata['x'] = torch.randn(10000000, 100)
sg = g.subgraph([0])
with open('a.pkl', 'wb') as f:
    pickle.dump(sg, f)     # Gives a pickle file of 4GB
sg.ndata['x']              # Reads the feature
with open('b.pkl', 'wb') as f:
    pickle.dump(sg, f)     # Gives a pickle file of 3KB
```

The reason is in the implementation of lazy feature slicing for node and edge features.

## Expected behavior

Pickle file of `sg` should take a small amount of space.

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
Fixes #2135 and #2137 by allowing the `storage` of `Column` to be `None`.

For subprocesses in PyTorch node and edge DataLoaders, we first check if the feature comes from the original graph by checking whether the column storage of the subgraph/block is identical to that of the original graph.  If so, we set the storage to be `None`.

After the subprocesses transmit the subgraphs/blocks, we then check in the main process if the column storage is None.  If so, we restore the storage to be that with the same name in the original graph.

### CODE REVIEW

File: python/dgl/dataloading/pytorch/__init__.py [hunk 1]
```diff
@@ -13,6 +13,129 @@ def _remove_kwargs_dist(kwargs):
         print('Distributed DataLoader does not support pin_memory')
     return kwargs
 
+# The following code is a fix to the PyTorch-specific issue in
+# https://github.com/dmlc/dgl/issues/2137
+#
```
Comment (jermainewang) [inline python/dgl/dataloading/pytorch/__init__.py:18]: ...

The sampled blocks/subgraphs contain features extracted from the parent graph. In DGL, the sub-feature
data structure holds a reference to the parent graph feature tensor and an index tensor. When transmitting
the sampled block/subgraph to the main process, it is unnecessary to transmit the parent graph feature
back.

To achieve that, in the collator ....

File: python/dgl/dataloading/pytorch/__init__.py [hunk 2]
```diff
@@ -13,6 +13,129 @@ def _remove_kwargs_dist(kwargs):
...
+        for etype in block.canonical_etypes:
+            subframe = block._edge_frames[block.get_etype_id(etype)]
+            frame = g._edge_frames[g.get_etype_id(etype)]
+            _restore_subframe_storage(subframe, frame)
+
+class _NodeCollator(NodeCollator):
+    def collate(self, items):
+        input_nodes, output_nodes, blocks = super().collate(items)
+        _pop_blocks_storage(blocks, self.g)
+        return input_nodes, output_nodes, blocks
```
Comment (jermainewang) [inline python/dgl/dataloading/pytorch/__init__.py:95]: Why not directly having `_pop_blocks_storage` in `NodeCollator`?

File: python/dgl/dataloading/pytorch/__init__.py [hunk 3]
```diff
@@ -13,6 +13,129 @@ def _remove_kwargs_dist(kwargs):
...
+            input_nodes, pair_graph, blocks = super().collate(items)
+            _pop_subgraph_storage(pair_graph, self.g)
+            _pop_blocks_storage(blocks, self.g_sampling)
+            return input_nodes, pair_graph, blocks
+        else:
+            input_nodes, pair_graph, neg_pair_graph, blocks = super().collate(items)
+            _pop_subgraph_storage(pair_graph, self.g)
+            _pop_subgraph_storage(neg_pair_graph, self.g)
+            _pop_blocks_storage(blocks, self.g_sampling)
+            return input_nodes, pair_graph, neg_pair_graph, blocks
```
Comment (jermainewang) [inline python/dgl/dataloading/pytorch/__init__.py:109]: same here. I feel the logic can go to `EdgeCollator` directly.

File: python/dgl/dataloading/pytorch/__init__.py [hunk 4]
```diff
@@ -13,6 +13,129 @@ def _remove_kwargs_dist(kwargs):
...
+            frame = g._node_frames[g.get_ntype_id(ntype)]
+            _pop_subframe_storage(subframe, frame)
+        for ntype in block.dsttypes:
+            subframe = block._node_frames[block.get_ntype_id_from_dst(ntype)]
+            frame = g._node_frames[g.get_ntype_id(ntype)]
+            _pop_subframe_storage(subframe, frame)
+        for etype in block.canonical_etypes:
+            subframe = block._edge_frames[block.get_etype_id(etype)]
+            frame = g._edge_frames[g.get_etype_id(etype)]
+            _pop_subframe_storage(subframe, frame)
```
Comment (jermainewang) [inline python/dgl/dataloading/pytorch/__init__.py:58]: What if the block/subgraph has new node/edge types that do not exist in the original graph?

File: python/dgl/dataloading/pytorch/__init__.py [hunk 5]
```diff
@@ -13,6 +13,129 @@ def _remove_kwargs_dist(kwargs):
...
+
+def _restore_subframe_storage(subframe, frame):
+    for key, col in subframe._columns.items():
+        if col.storage is None:
+            assert key in frame._columns
+            col.storage = frame._columns[key].storage
```
Comment (jermainewang) [inline python/dgl/dataloading/pytorch/__init__.py:64]: This check is only valid when a column's `storage` can never be `None`. Maybe add one sentence in `Column`'s docstring to emphasize that?

Comment (jermainewang) [reply python/dgl/dataloading/pytorch/__init__.py:64]: I feel the assertion at L63 is redundant, gives no extra information and incurs an extra key lookup. I'd suggest remove it.

File: python/dgl/dataloading/pytorch/__init__.py (hunk 2 above)
Comment (BarclayII) [reply python/dgl/dataloading/pytorch/__init__.py:95]: What if you write your own iterator that gets the blocks with NodeCollator (like in Tensorflow)?  The features of the blocks will be removed in this case.

File: python/dgl/dataloading/pytorch/__init__.py (hunk 5 above)
Comment (BarclayII) [reply python/dgl/dataloading/pytorch/__init__.py:64]: I left one paragraph in the docstring of `Column`.  Does that suffice?

File: python/dgl/dataloading/pytorch/__init__.py (hunk 1 above)
Comment (BarclayII) [reply python/dgl/dataloading/pytorch/__init__.py:18]: K.

File: python/dgl/dataloading/pytorch/__init__.py (hunk 4 above)
Comment (BarclayII) [reply python/dgl/dataloading/pytorch/__init__.py:77]: Right.  Added a check.

File: python/dgl/dataloading/pytorch/__init__.py (hunk 2 above)
Comment (BarclayII) [reply python/dgl/dataloading/pytorch/__init__.py:124]: Or we can put them in the node collator and have an argument `pop_storage` which defaults to False.

**Pull request merged**